constraints/miden-vm/hash.air

```diff
@@ -0,0 +1,243 @@
+mod HashChipletAir
+
+### Constants and periodic columns ################################################################
+
+periodic_columns:
+    cycle_row_7: [0, 0, 0, 0, 0, 0, 0, 1]
+    cycle_row_6: [0, 0, 0, 0, 0, 0, 1, 0]
+    cycle_row_0: [1, 0, 0, 0, 0, 0, 0, 0]
+
+
+### Helper functions ##############################################################################
+
+# Returns binary negation of the value.
+fn binary_not(value: scalar) -> scalar:
+    return 1 - value
+
+
+fn get_f_mp(s: vector[3]) -> scalar:
+    return cycle_row_0 * s[0] * binary_not(s[1]) * s[2]
+
+
+fn get_f_mv(s: vector[3]) -> scalar:
+    return cycle_row_0 * s[0] * s[1] * binary_not(s[2])
+
+
+fn get_f_mu(s: vector[3]) -> scalar:
+    return cycle_row_0 * s[0] * s[1] * s[2]
+
+
+fn get_f_abp(s: vector[3]) -> scalar:
+    return cycle_row_7 * s[0] * binary_not(s[1]) * binary_not(s[2])
+
+
+fn get_f_mpa(s: vector[3]) -> scalar:
+    return cycle_row_7 * s[0] * binary_not(s[1]) * s[2]
+
+
+fn get_f_mva(s: vector[3]) -> scalar:
+    return cycle_row_7 * s[0] * s[1] * binary_not(s[2])
+
+
+fn get_f_mua(s: vector[3]) -> scalar:
+    return cycle_row_7 * s[0] * s[1] * s[2]
+
+
+### Helper evaluators #############################################################################
+
+# Enforces that column must be binary.
+ev is_binary(main: [a]):
+    enf a^2 = a
+
+
+# Enforces that value in column is copied over to the next row.
+ev is_unchanged(main: [column]):
+    ev column' = column
+
+
+# Enforce selector columns constraints
+ev selector_columns(main: [s[3]]):
+    # We can define two flags:
+    # 1. Flag f_hout = cycle_row_7 * binary_not(s[0]) * binary_not(s[1]) * binary_not(s[2]), 
+    # which is set to 1 when selector flags are (0,0,0) on rows which are 1 less than a multiple 
+    # of 8. This is flag of the instruction that returns the result of the currently running 
+    # computation.
+    # 2. Flag f_sout = cycle_row_7 * binary_not(s[0]) * binary_not(s[1]) * s[2], which is set to 1 
+    # when selector flags are (0,0,1) on rows which are 1 less than a multiple of 8. This is flag 
+    # of the instruction that returns the whole hasher state.
+    #
+    # Flag f_out is set to 1 when either f_hout = 1 or f_sout = 1 in the current row.
+    let f_out = cycle_row_7 * binary_not(s[0]) * binary_not(s[1])
```

Comment:
Should we make a helper function for `f_out` as well? Then this comment can go above the function definition and this evaluator and the ones below can be much smaller